File: backend-scanner.py
#!/usr/bin/env python3
"""
GUNIC Advanced Security Scanner Backend
Scanner profissional sem limitações de CORS
"""

import requests
import socket
import ssl
import dns.resolver
import whois
import subprocess
import json
import time
import threading
from urllib.parse import urljoin, urlparse
from flask import Flask, request, jsonify
from flask_cors import CORS

app = Flask(__name__)
CORS(app)

class AdvancedSecurityScanner:
    def __init__(self):
        self.vulnerabilities = []
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        })
    
    def scan_target(self, url):
        """Scan completo sem limitações"""
        results = {
            'url': url,
            'timestamp': time.time(),
            'vulnerabilities': [],
            'open_ports': [],
            'subdomains': [],
            'technologies': {},
            'ssl_info': {},
            'dns_records': {},
            'whois_data': {}
        }
        
        domain = urlparse(url).netloc
        
        # 1. Port Scanning REAL
        results['open_ports'] = self.port_scan(domain)
        
        # 2. Subdomain Enumeration REAL  
        results['subdomains'] = self.subdomain_enum(domain)
        
        # 3. SSL/TLS Analysis REAL
        results['ssl_info'] = self.ssl_analysis(domain)
        
        # 4. DNS Enumeration REAL
        results['dns_records'] = self.dns_enum(domain)
        
        # 5. WHOIS Lookup REAL
        results['whois_data'] = self.whois_lookup(domain)
        
        # 6. Web Vulnerability Scanning REAL
        results['vulnerabilities'] = self.web_vuln_scan(url)
        
        # 7. Technology Detection REAL
        results['technologies'] = self.tech_detection(url)
        
        return results
    
    def port_scan(self, domain):
        """Port scanning real com socket"""
        print(f"[+] Scanning ports on {domain}")
        open_ports = []
        common_ports = [21, 22, 23, 25, 53, 80, 110, 143, 443, 993, 995, 3306, 3389, 5432, 6379, 27017]
        
        def scan_port(port):
            try:
                sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                sock.settimeout(2)
                result = sock.connect_ex((domain, port))
                if result == 0:
                    service = socket.getservbyport(port, 'tcp') if port < 1024 else 'unknown'
                    open_ports.append({'port': port, 'service': service})
                    print(f"[+] Port {port} open ({service})")
                sock.close()
            except:
                pass
        
        threads = []
        for port in common_ports:
            t = threading.Thread(target=scan_port, args=(port,))
            threads.append(t)
            t.start()
        
        for t in threads:
            t.join()
        
        return open_ports
    
    def subdomain_enum(self, domain):
        """Enumeração real de subdomínios"""
        print(f"[+] Enumerating subdomains for {domain}")
        subdomains = []
        wordlist = ['www', 'mail', 'ftp', 'admin', 'test', 'dev', 'staging', 'api', 'blog', 'shop', 'portal', 'secure', 'vpn', 'remote', 'backup', 'db', 'mysql', 'phpmyadmin', 'cpanel', 'webmail', 'support']
        
        def check_subdomain(sub):
            try:
                subdomain = f"{sub}.{domain}"
                socket.gethostbyname(subdomain)
                subdomains.append(subdomain)
                print(f"[+] Found subdomain: {subdomain}")
            except:
                pass
        
        threads = []
        for sub in wordlist:
            t = threading.Thread(target=check_subdomain, args=(sub,))
            threads.append(t)
            t.start()
        
        for t in threads:
            t.join()
        
        return subdomains
    
    def ssl_analysis(self, domain):
        """Análise SSL/TLS real"""
        print(f"[+] Analyzing SSL/TLS for {domain}")
        try:
            context = ssl.create_default_context()
            with socket.create_connection((domain, 443), timeout=10) as sock:
                with context.wrap_socket(sock, server_hostname=domain) as ssock:
                    cert = ssock.getpeercert()
                    return {
                        'version': ssock.version(),
                        'cipher': ssock.cipher(),
                        'cert_subject': dict(x[0] for x in cert['subject']),
                        'cert_issuer': dict(x[0] for x in cert['issuer']),
                        'cert_expires': cert['notAfter']
                    }
        except Exception as e:
            return {'error': str(e)}
    
    def dns_enum(self, domain):
        """Enumeração DNS real"""
        print(f"[+] DNS enumeration for {domain}")
        records = {}
        record_types = ['A', 'AAAA', 'MX', 'TXT', 'CNAME', 'NS', 'SOA']
        
        for record_type in record_types:
            try:
                answers = dns.resolver.resolve(domain, record_type)
                records[record_type] = [str(rdata) for rdata in answers]
            except:
                records[record_type] = []
        
        return records
    
    def whois_lookup(self, domain):
        """WHOIS lookup real"""
        print(f"[+] WHOIS lookup for {domain}")
        try:
            w = whois.whois(domain)
            return {
                'registrar': w.registrar,
                'creation_date': str(w.creation_date),
                'expiration_date': str(w.expiration_date),
                'name_servers': w.name_servers,
                'emails': w.emails
            }
        except Exception as e:
            return {'error': str(e)}
    
    def web_vuln_scan(self, url):
        """Scan de vulnerabilidades web real"""
        print(f"[+] Web vulnerability scanning for {url}")
        vulnerabilities = []
        
        # SQL Injection Test REAL
        sql_payloads = ["'", "' OR '1'='1", "' UNION SELECT 1,2,3--", "'; DROP TABLE users--"]
        params = ['id', 'user', 'search', 'q', 'page']
        
        for param in params:
            for payload in sql_payloads:
                try:
                    test_url = f"{url}?{param}={payload}"
                    response = self.session.get(test_url, timeout=5)
                    
                    # Detectar erros SQL
                    sql_errors = ['mysql_fetch', 'ORA-', 'Microsoft OLE DB', 'PostgreSQL', 'sqlite3.OperationalError']
                    for error in sql_errors:
                        if error.lower() in response.text.lower():
                            vulnerabilities.append({
                                'type': 'SQL Injection',
                                'severity': 'critical',
                                'url': test_url,
                                'description': f'SQL error detected with payload: {payload}'
                            })
                            print(f"[!] SQL Injection found: {param}")
                            break
                except:
                    pass
        
        # XSS Test REAL
        xss_payloads = ["<script>alert('XSS')</script>", "<img src=x onerror=alert('XSS')>"]
        for param in params:
            for payload in xss_payloads:
                try:
                    test_url = f"{url}?{param}={payload}"
                    response = self.session.get(test_url, timeout=5)
                    
                    if payload in response.text:
                        vulnerabilities.append({
                            'type': 'Cross-Site Scripting (XSS)',
                            'severity': 'high',
                            'url': test_url,
                            'description': f'XSS payload reflected: {payload}'
                        })
                        print(f"[!] XSS found: {param}")
                except:
                    pass
        
        # Directory Traversal REAL
        lfi_payloads = ["../../../etc/passwd", "..\\..\\..\\windows\\system32\\drivers\\etc\\hosts"]
        for payload in lfi_payloads:
            try:
                test_url = f"{url}?file={payload}"
                response = self.session.get(test_url, timeout=5)
                
                if 'root:' in response.text or 'localhost' in response.text:
                    vulnerabilities.append({
                        'type': 'Local File Inclusion (LFI)',
                        'severity': 'critical',
                        'url': test_url,
                        'description': f'File inclusion successful: {payload}'
                    })
                    print(f"[!] LFI found")
            except:
                pass
        
        return vulnerabilities
    
    def tech_detection(self, url):
        """Detecção de tecnologias real"""
        print(f"[+] Technology detection for {url}")
        try:
            response = self.session.get(url, timeout=10)
            headers = response.headers
            
            technologies = {
                'server': headers.get('Server', 'Unknown'),
                'powered_by': headers.get('X-Powered-By', 'Unknown'),
                'framework': 'Unknown',
                'cms': 'Unknown'
            }
            
            # Detectar CMS
            content = response.text.lower()
            if 'wp-content' in content:
                technologies['cms'] = 'WordPress'
            elif 'joomla' in content:
                technologies['cms'] = 'Joomla'
            elif 'drupal' in content:
                technologies['cms'] = 'Drupal'
            
            return technologies
        except Exception as e:
            return {'error': str(e)}

scanner = AdvancedSecurityScanner()

@app.route('/scan', methods=['POST'])
def scan_endpoint():
    data = request.json
    url = data.get('url')
    
    if not url:
        return jsonify({'error': 'URL required'}), 400
    
    try:
        results = scanner.scan_target(url)
        return jsonify(results)
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@app.route('/health', methods=['GET'])
def health_check():
    return jsonify({'status': 'Scanner backend online'})

if __name__ == '__main__':
    print("🚀 GUNIC Advanced Security Scanner Backend")
    print("⚠️  Use apenas com autorização!")
    print("🌐 Starting server on http://localhost:5000")
    app.run(host='0.0.0.0', port=5000, debug=True)

File: backend-scanner.js
#!/usr/bin/env node
/**
 * GUNIC Advanced Security Scanner Backend - JavaScript Version
 * Scanner profissional sem limitações de CORS
 */

const express = require('express');
const cors = require('cors');
const net = require('net');
const dns = require('dns').promises;
const https = require('https');
const http = require('http');
const { URL } = require('url');
const axios = require('axios');

const app = express();
app.use(cors());
app.use(express.json());

class AdvancedSecurityScanner {
    constructor() {
        this.vulnerabilities = [];
        this.session = axios.create({
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            },
            timeout: 10000
        });
    }

    async scanTarget(url) {
        console.log(`[+] Starting scan for: ${url}`);

        const results = {
            url: url,
            timestamp: Date.now(),
            vulnerabilities: [],
            open_ports: [],
            subdomains: [],
            technologies: {},
            ssl_info: {},
            dns_records: {},
            whois_data: {}
        };

        const domain = new URL(url).hostname;

        try {
            // 1. Port Scanning REAL
            results.open_ports = await this.portScan(domain);

            // 2. Subdomain Enumeration REAL
            results.subdomains = await this.subdomainEnum(domain);

            // 3. SSL/TLS Analysis REAL
            results.ssl_info = await this.sslAnalysis(domain);

            // 4. DNS Enumeration REAL
            results.dns_records = await this.dnsEnum(domain);

            // 5. WHOIS Lookup REAL
            results.whois_data = await this.whoisLookup(domain);

            // 6. Web Vulnerability Scanning REAL
            results.vulnerabilities = await this.webVulnScan(url);

            // 7. Technology Detection REAL
            results.technologies = await this.techDetection(url);

        } catch (error) {
            console.error(`[!] Error during scan: ${error.message}`);
        }

        return results;
    }

    async portScan(domain) {
        console.log(`[+] Scanning ports on ${domain}`);
        const openPorts = [];
        const commonPorts = [21, 22, 23, 25, 53, 80, 110, 143, 443, 993, 995, 3306, 3389, 5432, 6379, 27017];

        const scanPromises = commonPorts.map(port => this.scanPort(domain, port));
        const results = await Promise.allSettled(scanPromises);

        results.forEach((result, index) => {
            if (result.status === 'fulfilled' && result.value) {
                openPorts.push(result.value);
                console.log(`[+] Port ${commonPorts[index]} open (${result.value.service})`);
            }
        });

        return openPorts;
    }

    async scanPort(domain, port) {
        return new Promise((resolve) => {
            const socket = new net.Socket();
            socket.setTimeout(2000);

            socket.on('connect', () => {
                const service = this.getServiceName(port);
                socket.destroy();
                resolve({ port, service });
            });

            socket.on('timeout', () => {
                socket.destroy();
                resolve(null);
            });

            socket.on('error', () => {
                socket.destroy();
                resolve(null);
            });

            socket.connect(port, domain);
        });
    }

    getServiceName(port) {
        const services = {
            21: 'ftp', 22: 'ssh', 23: 'telnet', 25: 'smtp', 53: 'dns',
            80: 'http', 110: 'pop3', 143: 'imap', 443: 'https', 993: 'imaps',
            995: 'pop3s', 3306: 'mysql', 3389: 'rdp', 5432: 'postgresql',
            6379: 'redis', 27017: 'mongodb'
        };
        return services[port] || 'unknown';
    }

    async subdomainEnum(domain) {
        console.log(`[+] Enumerating subdomains for ${domain}`);
        const subdomains = [];
        const wordlist = ['www', 'mail', 'ftp', 'admin', 'test', 'dev', 'staging', 'api', 'blog', 'shop', 'portal', 'secure', 'vpn', 'remote', 'backup', 'db', 'mysql', 'phpmyadmin', 'cpanel', 'webmail', 'support'];

        const checkPromises = wordlist.map(sub => this.checkSubdomain(sub, domain));
        const results = await Promise.allSettled(checkPromises);

        results.forEach((result, index) => {
            if (result.status === 'fulfilled' && result.value) {
                subdomains.push(result.value);
                console.log(`[+] Found subdomain: ${result.value}`);
            }
        });

        return subdomains;
    }

    async checkSubdomain(sub, domain) {
        try {
            const subdomain = `${sub}.${domain}`;
            await dns.lookup(subdomain);
            return subdomain;
        } catch {
            return null;
        }
    }

    async sslAnalysis(domain) {
        console.log(`[+] Analyzing SSL/TLS for ${domain}`);
        try {
            return new Promise((resolve) => {
                const options = {
                    hostname: domain,
                    port: 443,
                    method: 'GET',
                    timeout: 10000
                };

                const req = https.request(options, (res) => {
                    const cert = res.socket.getPeerCertificate();
                    resolve({
                        version: res.socket.getProtocol(),
                        cipher: res.socket.getCipher(),
                        cert_subject: cert.subject,
                        cert_issuer: cert.issuer,
                        cert_expires: cert.valid_to
                    });
                });

                req.on('error', (error) => {
                    resolve({ error: error.message });
                });

                req.on('timeout', () => {
                    req.destroy();
                    resolve({ error: 'Timeout' });
                });

                req.end();
            });
        } catch (error) {
            return { error: error.message };
        }
    }

    async dnsEnum(domain) {
        console.log(`[+] DNS enumeration for ${domain}`);
        const records = {};
        const recordTypes = ['A', 'AAAA', 'MX', 'TXT', 'CNAME', 'NS', 'SOA'];

        for (const recordType of recordTypes) {
            try {
                const answers = await dns.resolve(domain, recordType);
                records[recordType] = answers;
            } catch {
                records[recordType] = [];
            }
        }

        return records;
    }

    async whoisLookup(domain) {
        console.log(`[+] WHOIS lookup for ${domain}`);
        try {
            // Usando API pública para WHOIS (alternativa ao módulo python-whois)
            const response = await axios.get(`https://whois.whoisxmlapi.com/api/v1?apiKey=demo&domainName=${domain}`, {
                timeout: 10000
            });

            const data = response.data;
            return {
                registrar: data.registrar?.name || 'Unknown',
                creation_date: data.creationDate || 'Unknown',
                expiration_date: data.expirationDate || 'Unknown',
                name_servers: data.nameServers || [],
                emails: data.emails || []
            };
        } catch (error) {
            return { error: error.message };
        }
    }

    async webVulnScan(url) {
        console.log(`[+] Web vulnerability scanning for ${url}`);
        const vulnerabilities = [];

        // SQL Injection Test REAL
        const sqlPayloads = ["'", "' OR '1'='1", "' UNION SELECT 1,2,3--", "'; DROP TABLE users--"];
        const params = ['id', 'user', 'search', 'q', 'page'];

        for (const param of params) {
            for (const payload of sqlPayloads) {
                try {
                    const testUrl = `${url}?${param}=${encodeURIComponent(payload)}`;
                    const response = await this.session.get(testUrl, { timeout: 5000 });

                    // Detectar erros SQL
                    const sqlErrors = ['mysql_fetch', 'ORA-', 'Microsoft OLE DB', 'PostgreSQL', 'sqlite3.OperationalError'];
                    for (const error of sqlErrors) {
                        if (response.data.toLowerCase().includes(error.toLowerCase())) {
                            vulnerabilities.push({
                                type: 'SQL Injection',
                                severity: 'critical',
                                url: testUrl,
                                description: `SQL error detected with payload: ${payload}`
                            });
                            console.log(`[!] SQL Injection found: ${param}`);
                            break;
                        }
                    }
                } catch {
                    // Ignora erros de conexão
                }
            }
        }

        // XSS Test REAL
        const xssPayloads = ["<script>alert('XSS')</script>", "<img src=x onerror=alert('XSS')>"];
        for (const param of params) {
            for (const payload of xssPayloads) {
                try {
                    const testUrl = `${url}?${param}=${encodeURIComponent(payload)}`;
                    const response = await this.session.get(testUrl, { timeout: 5000 });

                    if (response.data.includes(payload)) {
                        vulnerabilities.push({
                            type: 'Cross-Site Scripting (XSS)',
                            severity: 'high',
                            url: testUrl,
                            description: `XSS payload reflected: ${payload}`
                        });
                        console.log(`[!] XSS found: ${param}`);
                    }
                } catch {
                    // Ignora erros de conexão
                }
            }
        }

        // Directory Traversal REAL
        const lfiPayloads = ["../../../etc/passwd", "..\\..\\..\\windows\\system32\\drivers\\etc\\hosts"];
        for (const payload of lfiPayloads) {
            try {
                const testUrl = `${url}?file=${encodeURIComponent(payload)}`;
                const response = await this.session.get(testUrl, { timeout: 5000 });

                if (response.data.includes('root:') || response.data.includes('localhost')) {
                    vulnerabilities.push({
                        type: 'Local File Inclusion (LFI)',
                        severity: 'critical',
                        url: testUrl,
                        description: `File inclusion successful: ${payload}`
                    });
                    console.log(`[!] LFI found`);
                }
            } catch {
                // Ignora erros de conexão
            }
        }

        return vulnerabilities;
    }

    async techDetection(url) {
        console.log(`[+] Technology detection for ${url}`);
        try {
            const response = await this.session.get(url, { timeout: 10000 });
            const headers = response.headers;

            const technologies = {
                server: headers.server || 'Unknown',
                powered_by: headers['x-powered-by'] || 'Unknown',
                framework: 'Unknown',
                cms: 'Unknown'
            };

            // Detectar CMS
            const content = response.data.toLowerCase();
            if (content.includes('wp-content')) {
                technologies.cms = 'WordPress';
            } else if (content.includes('joomla')) {
                technologies.cms = 'Joomla';
            } else if (content.includes('drupal')) {
                technologies.cms = 'Drupal';
            }

            return technologies;
        } catch (error) {
            return { error: error.message };
        }
    }
}

const scanner = new AdvancedSecurityScanner();

// Rotas da API
app.post('/scan', async (req, res) => {
    const { url } = req.body;

    if (!url) {
        return res.status(400).json({ error: 'URL required' });
    }

    try {
        const results = await scanner.scanTarget(url);
        res.json(results);
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

app.get('/health', (req, res) => {
    res.json({ status: 'Scanner backend online' });
});

// Inicialização do servidor
const PORT = process.env.PORT || 5000;
const HOST = '0.0.0.0';

app.listen(PORT, HOST, () => {
    console.log("🚀 GUNIC Advanced Security Scanner Backend - JavaScript Version");
    console.log("⚠️  Use apenas com autorização!");
    console.log(`🌐 Starting server on http://localhost:${PORT}`);
});

module.exports = { AdvancedSecurityScanner, app };
